### ISSUE TITLE
Support valueless query parameters in the DSL

### ISSUE THREAD

**Issue by howyp** [http4s/http4s#2254]
I couldn't find a way to match valueless query parameters using the http4s DSL. 

These are the sort that are used as a flag ie:

`http://example.com/foo?bar`
vs 
`http://example.com/foo`

There’s no value for `bar`, but it’s presence or lack of indicates that the resource should behave differently. I tried an `OptionalQueryParamDecoderMatcher[Unit](“bar”)`, but that supplies `None` even if the parameter without a value - it returns `Some(())` only if an (potentially empty) value is supplied.

**Comment by howyp** [comment]
As was very helpfully suggested by @andimiller, this is a good workaround:

```scala
case req @ GET -> Root / blah if req.params.contains("foo") => ...
```

But I think it would be nice to find a way to do this directly in the DSL

**Comment by howyp** [comment]
Looking at the existing code, I guess we see the behaviour I described above because the `_.headOption` on the sequence of values occurs before the `QueryParamDecoder` is ever called:

```scala
abstract class OptionalQueryParamDecoderMatcher[T: QueryParamDecoder](name: String) {
  def unapply(params: Map[String, Seq[String]]): Option[Option[T]] =
    params
      .get(name)
      .flatMap(_.headOption)
      .traverse(s => QueryParamDecoder[T].decode(QueryParameterValue(s)))
      .toOption
}
```

So I guess if we want to fix this it may mean changing the API between the matchers and decoders ☹️

**Comment by rossabaker** [comment]
Fixed by #2266.
